# Lab 04 - pandas II

Name: Samuel Baldwin  
Class: CSCI 349 - Intro to Data Mining  
Section: 01 
Semester: Spring 2024
Instructors: Brian King, Joshua Stough


**Submission:**  lab04.ipynb  
**Points**: 10  
**Due**: Friday, February 2  
**Pair Programming:** OPTIONAL (welcome and encouraged!)  

# Objectives
-   Continuing from where lab03 left off, it's all about more experience with pandas

# Reminder – Pair Programming
If you are working with a partner, please make sure only one of you uploads the PDF to Gradescope and clearly choose both members in Gradescope so you both get credit. And, if you are working with a partner, please make sure you are following the pair programming guidelines, and be sure to put the names of both partners at the top of your submitted notebook!

# Introduction
As mentioned in the previous lab, a large portion of this lab is taken from snippets scattered throughout the enormous documentation and tutorials from the [pandas website](http://pandas.pydata.org/pandas-docs/stable/index.html).

Just a friendly reminder to not neglect the O'Reilly books referenced on Moodle. One book in particular, **Python for Data Analysis**, is written by the developer of pandas, Wes McKinney. This is one of our favorite reference books for using Python for Data Science.

We'll get you started with critical imports required for the lab:

In [448]:
import sys
import numpy as np
import pandas as pd

In [449]:
print(np.__version__)
print(pd.__version__)

1.26.3
2.1.4


---
# Exercises

Let's create a bigger version of the quiz scores dataset with the following features:

-   We have two students, Bob and Jane  
-   We have four weeks of data. Notice we can have duplicate names in our index. (However, columns must be unique! You can't have duplicate variable names!)  
-   There are missing quizzes, indicated with a zero in those locations.  


In [450]:
days = ["Mon","Tues","Wed","Thu","Fri","Student"]
scores = pd.DataFrame([pd.Series([8.75, 9.5, 8, 10, 7.75, "Bob"], index=days, name="week_1"),
                       pd.Series([8, 9, 10, 8.75, 7.25, "Jane"], index=days, name="week_1"),
                       pd.Series([0, 8, 9.75, 9, 6, "Bob"], index=days, name="week_2"),
                       pd.Series([8.25, 7, 0, 9.25, 8, "Jane"], index=days, name="week_2"),
                       pd.Series([8.5, 7, 9.25, 0, 0, "Bob"], index=days, name="week_3"),
                       pd.Series([8.25, 8.25, 0, 8, 7.5, "Jane"], index=days, name="week_3"),
                       pd.Series([6.5, 8.75, 8.5, 8, 6, "Bob"], index=days, name="week_4"),
                       pd.Series([10, 9.25, 8.5, 7.75, 7.5, "Jane"], index=days, name="week_4"),
                      ])
scores

Mon  Tues    Wed    Thu   Fri Student
week_1   8.75  9.50   8.00  10.00  7.75     Bob
week_1   8.00  9.00  10.00   8.75  7.25    Jane
week_2   0.00  8.00   9.75   9.00  6.00     Bob
week_2   8.25  7.00   0.00   9.25  8.00    Jane
week_3   8.50  7.00   9.25   0.00  0.00     Bob
week_3   8.25  8.25   0.00   8.00  7.50    Jane
week_4   6.50  8.75   8.50   8.00  6.00     Bob
week_4  10.00  9.25   8.50   7.75  7.50    Jane

Always start by learning a bit about your data. Generally, we look at the structure first, and then we explore it.

**1)**	 [P] Report the shape, the number of dimensions, the size of the data, and the data types of each column. Do this separately. Then create a separate cell that just shows the results of `scores.info()`.

In [451]:
print(f'Shape: {scores.shape}, Size: {scores.size}, Data type: {scores.dtypes}')

Shape: (8, 6), Size: 48, Data type: Mon        float64
Tues       float64
Wed        float64
Thu        float64
Fri        float64
Student     object
dtype: object


In [452]:
scores.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, week_1 to week_4
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Mon      8 non-null      float64
 1   Tues     8 non-null      float64
 2   Wed      8 non-null      float64
 3   Thu      8 non-null      float64
 4   Fri      8 non-null      float64
 5   Student  8 non-null      object 
dtypes: float64(5), object(1)
memory usage: 448.0+ bytes


**2)**	[P] Show two different ways to report the number of observations in scores.

In [453]:
scores.size

48

In [454]:
scores.shape[0] * scores.shape[1]

48

**3)**	[P] Show the variables (i.e. columns) in scores

In [455]:
scores.columns

Index(['Mon', 'Tues', 'Wed', 'Thu', 'Fri', 'Student'], dtype='object')

**4)**	[P] There are repetitive names in the index. This happens quite a bit, depending on how the data is organized. For now. print the unique index names in `scores`


In [456]:
scores.index.unique()

Index(['week_1', 'week_2', 'week_3', 'week_4'], dtype='object')

**5)**	[P] Rename the 'Tues' column header to be 'Tue' . (Hint: Changing `scores.columns[1]` directly will not work, not to mention it would be a poor approach to hard code a location this way! Magic numbers are never a good idea when referencing index or column locations. They may change!) Show the new scores data frame. This should be the same, except with Tues changed to Tue.



In [457]:
scores = scores.rename(columns = {'Tues':'Tue'})

**6)**	[P] Now, make a bigger change. Let's rename 'week_1' to be 'w1', 'week_2' to be 'w2', and so on. Also, suppose you decide to use only 2 letter abbreviations for the days. Rename the days in the column names to be 'Mo', 'Tu', 'We', 'Th', 'Fr'. Show the updated scores dataframe. It should appear something like the following:

```
       Mo    Tu     We     Th    Fr Student
w1   8.75  9.50   8.00  10.00  7.75     Bob
w1   8.00  9.00  10.00   8.75  7.25    Jane
w2   0.00  8.00   9.75   9.00  6.00     Bob
w2   8.25  7.00   0.00   9.25  8.00    Jane
w3   8.50  7.00   9.25   0.00  0.00     Bob
w3   8.25  8.25   0.00   8.00  7.50    Jane
w4   6.50  8.75   8.50   8.00  6.00     Bob
w4  10.00  9.25   8.50   7.75  7.50    Jane
```

In [459]:
rename = {"week_1":"w1","week_2":"w2","week_3":"w3","week_4":"w4","Mon":"Mo", "Tue":"Tu", "Wed":"We", "Thu":"Th", "Fri":"Fr"}
scores = scores.rename(index = rename, columns = rename)
scores

Mo    Tu     We     Th    Fr Student
w1   8.75  9.50   8.00  10.00  7.75     Bob
w1   8.00  9.00  10.00   8.75  7.25    Jane
w2   0.00  8.00   9.75   9.00  6.00     Bob
w2   8.25  7.00   0.00   9.25  8.00    Jane
w3   8.50  7.00   9.25   0.00  0.00     Bob
w3   8.25  8.25   0.00   8.00  7.50    Jane
w4   6.50  8.75   8.50   8.00  6.00     Bob
w4  10.00  9.25   8.50   7.75  7.50    Jane

**7)**	[M] Compare the type of the expression `scores['Mo']` vs. `scores[['Mo']]`. What is the difference? (HINT: Use the `type()` Python function on both.) Explain.

In [460]:
print(type(scores['Mo']))
print(scores['Mo'])
print(type(scores[['Mo']]))
print(scores[['Mo']])


<class 'pandas.core.series.Series'>
w1     8.75
w1     8.00
w2     0.00
w2     8.25
w3     8.50
w3     8.25
w4     6.50
w4    10.00
Name: Mo, dtype: float64
<class 'pandas.core.frame.DataFrame'>
       Mo
w1   8.75
w1   8.00
w2   0.00
w2   8.25
w3   8.50
w3   8.25
w4   6.50
w4  10.00


ANSWER: A single bracket around the search term of will return a Series containing just the data that is associated with the key. Double brackets will return a DataFrame that not only contains the data from the key, but will take thee slice of the original data frame, wit the key included.

**8)**	[P] Demonstrate the `describe()` method on scores. What type does it return? (This makes it quite usable!)

In [462]:
print(scores.describe())
type(scores.describe())

              Mo        Tu         We         Th        Fr
count   8.000000  8.000000   8.000000   8.000000  8.000000
mean    7.281250  8.343750   6.750000   7.593750  6.250000
std     3.095035  0.963045   4.219428   3.159276  2.635608
min     0.000000  7.000000   0.000000   0.000000  0.000000
25%     7.625000  7.750000   6.000000   7.937500  6.000000
50%     8.250000  8.500000   8.500000   8.375000  7.375000
75%     8.562500  9.062500   9.375000   9.062500  7.562500
max    10.000000  9.500000  10.000000  10.000000  8.000000


pandas.core.frame.DataFrame

ANSWER: The describe method returns a data frame that contains an extensive amount of information on the passed DataFrame, such as the 5 number summary for each key, the number of elements, standard deviation and mean.

**9)**	[P] Store the output of `describe()` as a variable. Then, using this data frame, report the day that had the largest standard deviation in quiz scores and its value. Clean your temporary variables when complete (i.e use `del`)

Your output should look something like:

`We had the highest standard deviation: 4.219`

In [471]:
describe = scores.describe()
print(f'{describe.iloc[2].idxmax()} had the highest standard deviation: {round(describe.iloc[2].max(), 3)}')
del(describe)

We had the highest standard deviation: 4.219


**10)**	[P] Knowing that `describe()` is nothing more than another data frame can be useful. However, it's more important to write code to output the specific results you are looking for. Write the Python code to repeat the previous exercise without using `describe()`.

In [474]:
stdev = scores.iloc[:, :-1].std().max()
print(f'{scores.iloc[:, :-1].std().idxmax()} had the highest standard deviation: {round(stdev,3)}')

We had the highest standard deviation: 4.219


---
OK. So, here's the point. LOOK AT THAT STANDARD DEVIATION! What is it telling you? If you've been doing everything correctly, then that result should be **\~4.219 which occurred on Wednesday**, and that is **way too large**!

**This is an important part of the preliminary EDA (Exploratory Data Analysis) step that helps you get a sense about the quality of your data.**

What does standard deviation measure? **It's a standardized quantity that tells you how much your data are deviating from the mean.* And, something is wrong with the data on Wednesday. Why? Think about it. Developing intuition with your data has immense value for the data scientist. This high s.d. value suggests some outliers or something else that is wrong. Notice, there are two missing values represented by 0 that are being used to compute the deviation, and that's not what we want. (You''ll notice there are other days with high values too, for similar reasons.) Missing data are usually not included with summary statistics (and many other calculations!) **We need to deal with missing data properly**.

When we have missing values in data, we almost always use a `na.nan` constant in its place to represent that the data is missing in that specific cell.  

Numpy has a variety of very useful constants. [Numpy's API reference on constants](https://numpy.org/doc/stable/reference/constants.html) has more information.  

---

**11)** [P] Write the code that changes all 0.0 entries to `np.nan`.  (NOTE: You should be able to do this with just one line of code using pandas selection techniques!) Then, show scores. All 0.0 entries should be replaced with NaN.

In [532]:
scores = scores.replace(0.0, np.nan)
print(scores)

       Mo    Tu     We     Th    Fr Student
w1   8.75  9.50   8.00  10.00  7.75     Bob
w1   8.00  9.00  10.00   8.75  7.25    Jane
w2    NaN  8.00   9.75   9.00  6.00     Bob
w2   8.25  7.00    NaN   9.25  8.00    Jane
w3   8.50  7.00   9.25    NaN   NaN     Bob
w3   8.25  8.25    NaN   8.00  7.50    Jane
w4   6.50  8.75   8.50   8.00  6.00     Bob
w4  10.00  9.25   8.50   7.75  7.50    Jane


## Important 

**You must always pay attention to NaN values in your data. These represent missing values and are important to ignore in most calculations, especially when you are conducting your preliminary EDA. We will learn about many techniques for dealing with missing data soon. Many functions have an optional parameter such as skipna to make sure that whatever you are trying to do does not include these missing data. Pay attention to those parameters. Most of the time, the default parameter values for handling NaN values is to skip or ignore them.**


**12)**	[P] Show the output of `describe()` again. You should notice much more reasonable SD values. Note how much more informative this output is when properly dealing with missing data. Then, copy your code from above to output the day with the largest standard deviation in quiz scores. You should have a substantially different result than you did before.

In [484]:
print(scores.describe())
stdev = scores.iloc[:, :-1].std().max()
print(f'We had the highest standard deviation: {round(stdev,3)}')

              Mo        Tu         We         Th        Fr
count   7.000000  8.000000   6.000000   7.000000  7.000000
mean    8.321429  8.343750   9.000000   8.678571  7.142857
std     1.037970  0.963045   0.790569   0.812843  0.814672
min     6.500000  7.000000   8.000000   7.750000  6.000000
25%     8.125000  7.750000   8.500000   8.000000  6.625000
50%     8.250000  8.500000   8.875000   8.750000  7.500000
75%     8.625000  9.062500   9.625000   9.125000  7.625000
max    10.000000  9.500000  10.000000  10.000000  8.000000
We had the highest standard deviation: 1.038


**13)**	[PM] Show `scores.values`. What does the `.values` attribute do?

In [485]:
scores.values

array([[8.75, 9.5, 8.0, 10.0, 7.75, 'Bob'],
       [8.0, 9.0, 10.0, 8.75, 7.25, 'Jane'],
       [nan, 8.0, 9.75, 9.0, 6.0, 'Bob'],
       [8.25, 7.0, nan, 9.25, 8.0, 'Jane'],
       [8.5, 7.0, 9.25, nan, nan, 'Bob'],
       [8.25, 8.25, nan, 8.0, 7.5, 'Jane'],
       [6.5, 8.75, 8.5, 8.0, 6.0, 'Bob'],
       [10.0, 9.25, 8.5, 7.75, 7.5, 'Jane']], dtype=object)

ANSWER: The .values attribute shows the columns of the DataFrame, in this case, giving each score for each person within the Frame

The `.values` attribute of a DataFrame returns a basic numpy array

**14)**	[P] What is the mean for each day? Do not use `describe()`. Round your result to 3 significant digits.

In [486]:
np.round(np.nanmean(scores.iloc[:, :-1].values, axis = 0),3)

array([8.321, 8.344, 9.   , 8.679, 7.143])

**15)**	[P] What is the mean quiz score over the entire dataset, ignoring all missing values? (HINT: there are many ways to do this, and no solution should involve writing any loops! **The only right answer is ~8.2786.** If you aren't getting that, then you are not handling NaN values correctly. You only want the mean over all scores that are NOT NaN!)

In [527]:
np.round(np.nanmean(scores.iloc[:, :-1].values.flatten()),4)

8.2786

**16)** [P] Show the mean for each week over all students. (Yes, technically you could write a loop to compute this exhaustively. However, you should rarely need to do that. Use pandas!)

HINT: You have the correct result if you report the following
```
Mean score for each week over all students
w1    8.700000
w2    8.156250
w3    8.107143
w4    8.075000
```

In [531]:
np.nanmean(scores.iloc[:, :-1].groupby(level=0).mean(), axis=1)

array([8.7  , 8.325, 8.15 , 8.075])

**17)** [P] Show the mean score for each student. Again, ignore all missing values.

HINT: You have the correct result if these are your results:
```
Mean score for each student
Student
Bob     8.191176
Jane    8.361111
``` 

In [491]:
temp = scores.groupby('Student').apply(lambda group: np.nanmean(group.values))
print(f'Mean score for each student\n {temp}')
del(temp)

Mean score for each student
 Student
Bob     8.191176
Jane    8.361111
dtype: float64


**18)**	[P] Select the data frame for only Monday and Student variables using `.loc`, and again using `.iloc`

In [492]:
# Using .loc
scores.loc[:, ['Mo', 'Student']]

Mo Student
w1   8.75     Bob
w1   8.00    Jane
w2    NaN     Bob
w2   8.25    Jane
w3   8.50     Bob
w3   8.25    Jane
w4   6.50     Bob
w4  10.00    Jane

In [493]:
# Using .iloc
scores.iloc[:,[0,-1]]

Mo Student
w1   8.75     Bob
w1   8.00    Jane
w2    NaN     Bob
w2   8.25    Jane
w3   8.50     Bob
w3   8.25    Jane
w4   6.50     Bob
w4  10.00    Jane

**19)**	[P] Show a new DataFrame containing the week as an index (like the original), but the mean of their scores and the count of the quizzes they took. (HINT: Sometimes, it's just easier to create a new data frame.)

You should have results that look like the following:
```
   Student  count    mean
w1     Bob      5  8.8000
w1    Jane      5  8.6000
w2     Bob      4  8.1875
w2    Jane      4  8.1250
w3     Bob      3  8.2500
w3    Jane      4  8.0000
w4     Bob      5  7.5500
w4    Jane      5  8.6000
```


In [494]:
temp = pd.DataFrame({"Students":scores.iloc[:, -1],"Count": scores.count(axis=1)-1, "Mean":scores.iloc[:, :-1].mean(axis=1)})
temp.index = scores.index
print(temp)
del(temp)

   Students  Count    Mean
w1      Bob      5  8.8000
w1     Jane      5  8.6000
w2      Bob      4  8.1875
w2     Jane      4  8.1250
w3      Bob      3  8.2500
w3     Jane      4  8.0000
w4      Bob      5  7.5500
w4     Jane      5  8.6000


**20)**	[P] This time, show the total number of quizzes taken each week by all students. 

Your results should look something like the following:
```
w1    10
w2     8
w3     7
w4    10
```

In [495]:
temp = scores.groupby(level=0).count().sum(axis=1) - 2 
print(temp)
del(temp)

w1    10
w2     8
w3     7
w4    10
dtype: int64


**21)** [P] Show the number of times Friday's score was greater than 7.25 for each student. HINT: Start by selecting a subset of observations, then use value_counts() on the Student variable of the selected data

In [496]:
scores[scores['Fr'] > 7.25].groupby('Student').size()

Student
Bob     1
Jane    3
dtype: int64

**22)**	[P] Repeat the previous exercise, but this time use the `where()` method on scores to select your data. You should have the same output.

In [497]:
scores.groupby('Student')['Fr'].apply(lambda x: x.where(x > 7.25).count())

Student
Bob     1
Jane    3
Name: Fr, dtype: int64

**23)**	[P] Select the scores that were greater than the mean score over the entire dataset (without the missing values). Again, no hard coding values here (i.e. do NOT hard code the mean, but recompute it. Then, report the number of scores for each day that exceeded that global mean. Your result should look like

```
Mo    3
Tu    4
We    5
Th    4
Fr    0
```

In [498]:
mean = np.round(np.nanmean(scores.iloc[:, :-1].values.flatten()),4)
num = (scores.iloc[:,:-1] > mean).sum()
print(num)
del(mean)
del(num)

Mo    3
Tu    4
We    5
Th    4
Fr    0
dtype: int64


**24)**	[P] Sometimes when we deal with enormous datasets for modeling, we drop all observations that have any missing data. Show a new data frame that has only complete observations (rows). HINT: This is actually very simple to do. Check out the `dropna()` method.

In [499]:
scores.dropna()

Mo    Tu    We     Th    Fr Student
w1   8.75  9.50   8.0  10.00  7.75     Bob
w1   8.00  9.00  10.0   8.75  7.25    Jane
w4   6.50  8.75   8.5   8.00  6.00     Bob
w4  10.00  9.25   8.5   7.75  7.50    Jane

**25)** [P] Let's assume that missing quizzes implies that the student earns a zero for that score. Show a dataframe that reports the average quiz score for each student and each week with this assumption. Your result should be organized such that rows are the week, and columns are the student. 

Warning: Do NOT destroy the original `scores` dataframe that you cleaned! Leave the NaN values in there alone. 

You should have a result that looks like the following:

```
Week      w1    w2    w3    w4
Student                       
Bob      8.8  6.55  4.95  7.55
Jane     8.6  6.50  6.40  8.60
```



In [500]:
temp = scores.replace(np.nan, 0)
temp.groupby('Student').apply(lambda x: x.iloc[:, :-1].groupby(level=0).mean().mean(axis=1))

w1    w2    w3    w4
Student                       
Bob      8.8  6.55  4.95  7.55
Jane     8.6  6.50  6.40  8.60

**26)** [P] It is quite common that we drop the lowest score for students before computing our stats. Repeat the previous exercise, but remove the lowest score for each student in each week before computing the mean. If a week has NaN, then that is their lowest score.

Your result should look like the following:

```
Week       w1    w2    w3    w4
Student                        
Bob      9.06  8.19  6.19  7.94
Jane     8.94  8.12  8.00  8.88
```


In [501]:
temp = scores.replace(np.nan, 0)
temp = temp.groupby('Student').apply(lambda x: x.iloc[:, :-1].apply(lambda y: y.drop(y.idxmin()), axis=1).groupby(level=0).mean().mean(axis=1))
np.round(temp, 2)


w1    w2    w3    w4
Student                        
Bob      9.06  8.19  6.19  7.94
Jane     8.94  8.12  8.00  8.88